File: src/main/java/no/novari/flyt/authorization/client/sourceapplications/SourceApplicationController.java
package no.novari.flyt.authorization.client.sourceapplications;

import lombok.RequiredArgsConstructor;
import no.novari.flyt.authorization.client.sourceapplications.model.SourceApplication;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Comparator;
import java.util.List;

import static no.novari.flyt.resourceserver.UrlPaths.INTERNAL_API;

@RestController
@RequestMapping(INTERNAL_API + "/authorization/sourceapplications")
@RequiredArgsConstructor
public class SourceApplicationController {

    private final List<SourceApplication> sourceApplications;

    @GetMapping
    public ResponseEntity<List<SourceApplicationResponse>> getAll() {
        List<SourceApplicationResponse> response = sourceApplications.stream()
                .map(sourceApplication -> new SourceApplicationResponse(
                        sourceApplication.getSourceApplicationId(),
                        sourceApplication.getDisplayName()
                ))
                .sorted(Comparator.comparing(SourceApplicationResponse::displayName, String.CASE_INSENSITIVE_ORDER))
                .toList();
        return ResponseEntity.ok(response);
    }

    public record SourceApplicationResponse(long sourceApplicationId, String displayName) {
    }
}

File: src/main/kotlin/no/novari/flyt/authorization/client/sourceapplications/SourceApplicationController.kt
package no.novari.flyt.authorization.client.sourceapplications

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import no.novari.flyt.authorization.client.sourceapplications.model.SourceApplication
import no.novari.flyt.webresourceserver.UrlPaths.INTERNAL_API
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("$INTERNAL_API/authorization/sourceapplications")
@Tag(name = "Source applications", description = "Source applications available to Flyt users.")
class SourceApplicationController(
    private val sourceApplications: List<SourceApplication>,
) {
    @GetMapping
    @Operation(summary = "List source applications")
    fun getAll(): List<SourceApplicationResponse> {
        return sourceApplications
            .map { sourceApplication ->
                SourceApplicationResponse(
                    id = sourceApplication.id,
                    displayName = sourceApplication.displayName,
                    available = sourceApplication.available,
                )
            }.sortedBy { it.displayName.lowercase() }
    }

    data class SourceApplicationResponse(
        val id: Long,
        val displayName: String,
        val available: Boolean,
    )
}
